sbi/inference/posteriors/score_posterior.py

```diff
@@ -0,0 +1,393 @@
+
+        theta = ensure_theta_batched(torch.as_tensor(theta))
+        theta_density_estimator = reshape_to_sample_batch_event(
+            theta, theta.shape[1:], leading_is_sample=True
+        )
+        x_density_estimator = reshape_to_batch_event(
+            x, event_shape=self.score_estimator.condition_shape
+        )
+        assert (
+            x_density_estimator.shape[0] == 1
+        ), ".log_prob() supports only `batchsize == 1`."
+
+        self.score_estimator.eval()
+
+        # Compute the base density
+        mean_T = self.score_estimator.mean_T
+        std_T = self.score_estimator.std_T
+        base_density = torch.distributions.Normal(mean_T, std_T)
+        for _ in range(len(self.score_estimator.input_shape)):
+            base_density = torch.distributions.Independent(base_density, 1)
+        # Build the freeform jacobian transformation by probability flow ODEs
+        transform = self.build_freeform_jacobian_transform(
+            x_density_estimator, atol=atol, rtol=rtol, exact=exact
+        )
+
+        with torch.set_grad_enabled(track_gradients):
+            eps_samples, logabsdet = transform.inv.call_and_ladj(  # type: ignore
+                theta_density_estimator
+            )
+            base_log_prob = base_density.log_prob(eps_samples)
+            log_probs = base_log_prob - logabsdet
+            log_probs = log_probs.squeeze(-1)
+
+            # Force probability to be zero outside prior support.
+            in_prior_support = within_support(self.prior, theta)
+
+            masked_log_prob = torch.where(
+                in_prior_support,
+                log_probs,
+                torch.tensor(float("-inf"), dtype=torch.float32, device=self._device),
+            )
+
+            return masked_log_prob
+
+    def sample_batched(
+        self,
+        sample_shape: torch.Size,
+        x: Tensor,
+        max_sampling_batch_size: int = 10000,
+        show_progress_bars: bool = True,
+    ) -> Tensor:
+        raise NotImplementedError(
+            "Batched sampling is not implemented for ScorePosterior."
+        )
+
+    def map(
+        self,
+        x: Optional[Tensor] = None,
+        num_iter: int = 1000,
+        num_to_optimize: int = 100,
+        learning_rate: float = 1e54,
+        init_method: Union[str, Tensor] = "posterior",
+        num_init_samples: int = 1000,
+        save_best_every: int = 10,
+        show_progress_bars: bool = False,
+        force_update: bool = False,
+    ) -> Tensor:
+        r"""Returns the maximum-a-posteriori estimate (MAP).
+
+        The MAP is obtained by running gradient
+        ascent from a given number of starting positions (samples from the posterior
+        with the highest log-probability). After the optimization is done, we select the
+        parameter set that has the highest log-probability after the optimization.
+
+         Args:
+             x: Deprecated - use `.set_default_x()` prior to `.map()`.
```

Comment:
As all the score estimation API is new, we can remove the deprecated arguments, warnings, etc.?

Reply:
I will keep it for consistency with the other methods